hoomd/device.py

```diff
@@ -0,0 +1,381 @@
+        many times during a job script.
+
+        """
+
+        self.cpp_msg.setNoticeLevel(notice_level)
+
+    @property
+    def memory_traceback():
+        
+        return self.cpp_exec_conf.getMemoryTracer() is not None
+        
+    @memory_traceback.setter
+    def memory_traceback(self, mem_traceback):
+        
+        self.cpp_exec_conf.setMemoryTracing(mem_traceback)
+
+    @property
+    def gpu_error_checking(self):
+        
+        if self.mode == 'gpu':
+            return self.cpp_exec_conf.isCUDAErrorCheckingEnabled()
+        else:
+            self.cpp_msg.warning("Attempting to access gpu_error_checking while HOOMD is in CPU mode, returning False.\n")
+            return False
+    
+    @gpu_error_checking.setter
+    def gpu_error_checking(self, new_bool):
+        
+        if self.mode == 'gpu':
+            self.cpp_exec_conf.setCUDAErrorChecking(new_bool)
+        else:
+            self.cpp_msg.warning("HOOMD is in CPU mode, ignoring request to set gpu_error_checking.\n")
+
+    def set_msg_file(self, fname):
+        R""" Set the message file.
+
+        Args:
+            fname (str): Specifies the name of the file to write. The file will be overwritten.
+                         Set to None to direct messages back to stdout/stderr.
+
+        The message file may be changed before or after initialization, and may be changed many times during a job script.
+        Changing the message file will only affect messages sent after the change.
+
+        """
+
+        if fname is not None:
+            self.cpp_msg.openFile(fname)
+        else:
+            self.cpp_msg.openStd()
+
+
+## Initializes the Messenger
+# \internal
+def _create_messenger(mpi_config, notice_level, msg_file, shared_msg_file):
+    global _cpp_msgs
+
+    msg = _hoomd.Messenger(mpi_config)
+
+    # try to detect if we're running inside an MPI job
+    inside_mpi_job = mpi_config.getNRanksGlobal() > 1
+    if ('OMPI_COMM_WORLD_RANK' in os.environ or
+        'MV2_COMM_WORLD_LOCAL_RANK' in os.environ or
+        'PMI_RANK' in os.environ or
+        'ALPS_APP_PE' in os.environ):
+        inside_mpi_job = True
+
+    # only open python stdout/stderr in non-MPI runs
+    if not inside_mpi_job:
+        msg.openPython();
+
+    if notice_level is not None:
+        msg.setNoticeLevel(notice_level);
+
+    if msg_file is not None:
+        msg.openFile(msg_file);
+
+    if shared_msg_file is not None:
+        if not _hoomd.is_MPI_available():
+            msg.error("Shared log files are only available in MPI builds.\n");
+            raise RuntimeError('Error setting option');
+        msg.setSharedFile(shared_msg_file);
+
+    # add this messenger to the global list of messengers
+    _cpp_msgs.append(msg)
+
+    return msg
+
+
+def _init_nthreads(nthreads):
+    """
+    initializes the number of threads
+    """
+    
+    if nthreads is not None:
+        self.num_threads = nthreads
```

Comment:
Is it valid to pass `None` here? `None` is the default argument for `nthreads` in `CPU` and `Auto` below.